Validation rework

sungam3r opened the pull request:
The main goal of this PR is reduce memory allocations during validation. The basic idea is that `ValidationContext` stops being captured in closures. Visitors generated by validation rules now store their state in `ValidationContext`.

Also fixes benchmarks.

Results before:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
|        Method |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
| Introspection | 5,080.81 μs | 58.344 μs | 54.575 μs |      - |     - |     - | 863.53 KB |
|          Hero |    70.00 μs |  0.136 μs |  0.121 μs | 5.3711 |     - |     - |  22.04 KB |

Results after:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
|        Method |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
| Introspection | 5,240.58 μs | 73.260 μs | 68.527 μs |      - |     - |     - | 854.05 KB |
|          Hero |    68.13 μs |  0.035 μs |  0.029 μs | 3.0518 |     - |     - |  12.55 KB |

sungam3r commented:
@Shane32 I had to resolve conflicts twice on ~30 files.

sungam3r commented on src/Directory.Build.props:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <VersionPrefix>4.0.0-preview</VersionPrefix>
-    <LangVersion>latest</LangVersion>
+    <LangVersion>8.0</LangVersion>
```
Required by Benchmarks.Dotnet (only version number, not alias).

sungam3r commented on src/GraphQL/Validation/Rules/ArgumentsOfCorrectType.cs:
```diff
@@ -18,11 +18,7 @@ public class ArgumentsOfCorrectType : IValidationRule
         public static readonly ArgumentsOfCorrectType Instance = new ArgumentsOfCorrectType();
 
-        /// <inheritdoc/>
-        /// <exception cref="ArgumentsOfCorrectTypeError"/>
```
Generally speaking, this does not correspond to reality, because all such methods do not throw mentioned exceptions.

Shane32 replied:
I know, but I thought it was useful.

sungam3r replied:
agree

sungam3r commented:
The introspection execution time became longer. It may have been influenced by applications on my machine.

Shane32 commented:
> The introspection execution time became longer. It may have been influenced by applications on my machine.

No, it's the dictionary access.  You are trading a very few heap allocations for many lookups within a dictionary during execution.  It's killing the performance.

Shane32 commented:
> @Shane32 I had to resolve conflicts twice on ~30 files.

Well, I did explicitly say that I was going to put in a PR for work on the validation code, and asked you to wait until I had done so.  You wasted a lot of effort.

Shane32 commented:
> > The introspection execution time became longer. It may have been influenced by applications on my machine.
> 
> No, it's the dictionary access. You are trading a very few heap allocations for many lookups within a dictionary during execution. It's killing the performance.

I'm going to run some performance tests to be sure this is accurate.  It should be very easy to run repeatable performance tests against the original code, my PR, and your PR.

Shane32 commented:
> > The introspection execution time became longer. It may have been influenced by applications on my machine.
> 
> No, it's the dictionary access. You are trading a very few heap allocations for many lookups within a dictionary during execution. It's killing the performance.

Another thing: the dictionary is of type object so any value data stored in the dictionary needs to be boxed.  So the `int` in the `LoneAnonymousOperation` rule gets boxed (an allocation), and the `UniqueInputFieldNames.Names` class is a new class requiring an allocation.

Shane32 commented:
I'm going to move your benchmark modifications to a new PR so we figure this out easier, ok?

sungam3r commented:
> I'm going to move your benchmark modifications to a new PR so we figure this out easier, ok?

sungam3r commented:
>  So the int in the LoneAnonymousOperation rule gets boxed (an allocation)

sungam3r commented:
Regarding dictionary lookups - of course, this is slower compared to what it was. Here we can think further.

sungam3r commented:
> Well, I did explicitly say that I was going to put in a PR for work on the validation code, and asked you to wait until I had done so. You wasted a lot of effort.

By then I was almost finished.

Shane32 commented:
My benchmarks:
```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.685 (2004/?/20H1)
Intel Core i7-9700K CPU 3.60GHz (Coffee Lake), 1 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.1.301
  [Host]     : .NET Core 3.1.5 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.27001), X64 RyuJIT
  DefaultJob : .NET Core 3.1.5 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.27001), X64 RyuJIT



*****master*****

|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,232.29 us | 4.817 us | 4.506 us | 123.0469 | 42.9688 |     - |  754.6 KB |
|          Hero |    30.57 us | 0.195 us | 0.182 us |   4.0894 |  0.1221 |     - |  25.16 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,225.20 us | 6.093 us | 5.088 us | 123.0469 | 42.9688 |     - |  754.6 KB |
|          Hero |    30.53 us | 0.195 us | 0.173 us |   4.0894 |  0.1221 |     - |  25.16 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,231.13 us | 5.271 us | 4.672 us | 123.0469 | 42.9688 |     - |  754.6 KB |
|          Hero |    30.70 us | 0.152 us | 0.142 us |   4.0894 |  0.1221 |     - |  25.16 KB |


*****develop*****


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,168.20 us | 4.745 us | 3.962 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.10 us | 0.098 us | 0.092 us |   3.5706 |  0.1221 |     - |  22.04 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,171.40 us | 4.869 us | 4.316 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.44 us | 0.125 us | 0.117 us |   3.5706 |  0.0916 |     - |  22.04 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,160.41 us | 4.980 us | 4.658 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.07 us | 0.062 us | 0.058 us |   3.5706 |  0.0916 |     - |  22.04 KB |


*****validation_cleanup******


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,184.31 us | 4.083 us | 3.819 us | 117.1875 | 35.1563 |     - | 726.55 KB |
|          Hero |    26.21 us | 0.082 us | 0.069 us |   2.0447 |  0.0305 |     - |  12.55 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,188.03 us | 5.921 us | 5.538 us | 117.1875 | 35.1563 |     - | 726.55 KB |
|          Hero |    26.52 us | 0.127 us | 0.119 us |   2.0447 |       - |     - |  12.55 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,184.41 us | 4.754 us | 3.712 us | 117.1875 | 37.1094 |     - | 726.55 KB |
|          Hero |    26.55 us | 0.081 us | 0.076 us |   2.0447 |  0.0305 |     - |  12.55 KB |


*****validation_improvements*****


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,157.07 us | 5.014 us | 4.691 us | 119.1406 | 39.0625 |     - | 730.26 KB |
|          Hero |    26.55 us | 0.109 us | 0.097 us |   2.8076 |  0.0305 |     - |  17.22 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,152.70 us | 4.604 us | 4.306 us | 119.1406 | 37.1094 |     - | 730.26 KB |
|          Hero |    26.52 us | 0.068 us | 0.060 us |   2.8076 |  0.0305 |     - |  17.22 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,149.94 us | 4.578 us | 4.282 us | 119.1406 | 37.1094 |     - | 730.26 KB |
|          Hero |    26.32 us | 0.125 us | 0.117 us |   2.8076 |  0.0610 |     - |  17.22 KB |
```

sungam3r commented:
Where is what?

sungam3r commented:
That is, what are these three tables in each group?

sungam3r commented on src/GraphQL/Validation/Rules/LoneAnonymousOperation.cs:
```diff
@@ -17,20 +17,21 @@ public class LoneAnonymousOperation : IValidationRule
                 {
-                    if (string.IsNullOrWhiteSpace(op.Name)
-                        && operationCount > 1)
+                    int operationCount = context.Get<LoneAnonymousOperation, int>();
```
🤦 I could just use `context.Document.Operations.Count` directly.

Shane32 replied:
Take a look at what I did in my PR.  I short-circuited the entire operation.

sungam3r replied:
Already looking into. What do you mean _short-circuited_ (LoneAnonymousOperation.cs )?

Shane32 replied:
If you looked at the code, I assume you figured it out.  But basically, it just returns a "null" INodeVisitor after performing the context-wide check short-circuit check.

sungam3r replied:
I saw this in another file, not this one.

Shane32 commented:
> That is, what are these three tables in each group?
> 
> In any case, I think that relatively short requests are sent more often and we need to focus more on them.

I ran it three times for good measure.

Shane32 commented:
> In any case, I think that relatively short requests are sent more often and we need to focus more on them.

I have a complete e-commerce website running and the majority of the calls are not "short requests", at least, not very short.  I guess it depends on what you mean by "short" and the application's use case.

sungam3r commented:
Yeah. I've already changed my mind. The request in practice can be really big.

sungam3r commented:
@Shane32 I pushed new commit with further improvements. I have completely looked at your PR to make sure I haven't missed anything. The only place that made me doubt was `public INode[] GetAncestors()`.

sungam3r commented:
I think it should work even faster. Allocations will definitely become even less.

sungam3r commented:
Nope, only allocations less.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
|        Method |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
| Introspection | 5,407.21 μs | 66.088 μs | 61.819 μs |      - |     - |     - | 853.26 KB |
|          Hero |    71.70 μs |  0.062 μs |  0.058 μs | 2.8076 |     - |     - |  11.81 KB |

Shane32 commented:
Ok I just pushed a lot of improvements to my branch.  Going to run perf tests now.

Shane32 commented:
I have not looked at your code yet (and likely as not you've already done it), but we can move the state variables for the various rules into TypeInfo.  I've not done that yet, but it would certainly be better.

Shane32 commented:
Something is wrong with those changes.  It's running a lot slower than it was before, which was already slower than `develop`.

sungam3r commented:
Try on your machine.

Shane32 commented:
My changes improved both allocations and performance.  It's now running a full 10% faster than your branch for an introspection query, and with less allocations.

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.685 (2004/?/20H1)
Intel Core i7-9700K CPU 3.60GHz (Coffee Lake), 1 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.1.301
  [Host]     : .NET Core 3.1.5 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.27001), X64 RyuJIT
  DefaultJob : .NET Core 3.1.5 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.27001), X64 RyuJIT


*****develop*****


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,168.20 us | 4.745 us | 3.962 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.10 us | 0.098 us | 0.092 us |   3.5706 |  0.1221 |     - |  22.04 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,171.40 us | 4.869 us | 4.316 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.44 us | 0.125 us | 0.117 us |   3.5706 |  0.0916 |     - |  22.04 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,160.41 us | 4.980 us | 4.658 us | 119.1406 | 37.1094 |     - | 736.04 KB |
|          Hero |    28.07 us | 0.062 us | 0.058 us |   3.5706 |  0.0916 |     - |  22.04 KB |


*****validation_cleanup******


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,228.00 us | 3.018 us | 2.675 us | 117.1875 | 37.1094 |     - | 725.77 KB |
|          Hero |    26.33 us | 0.136 us | 0.114 us |   1.9226 |       - |     - |  11.81 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,229.17 us | 4.568 us | 4.049 us | 117.1875 | 37.1094 |     - | 725.76 KB |
|          Hero |    26.31 us | 0.139 us | 0.130 us |   1.9226 |       - |     - |  11.81 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,230.01 us | 5.698 us | 5.330 us | 117.1875 | 37.1094 |     - | 725.76 KB |
|          Hero |    26.59 us | 0.130 us | 0.122 us |   1.9226 |       - |     - |  11.81 KB |


*****validation_improvements*****


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,128.08 us | 6.333 us | 5.924 us | 117.1875 | 35.1563 |     - | 729.26 KB |
|          Hero |    24.48 us | 0.077 us | 0.068 us |   2.6245 |  0.0305 |     - |  16.24 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,123.21 us | 8.194 us | 7.665 us | 117.1875 | 35.1563 |     - | 729.26 KB |
|          Hero |    24.86 us | 0.216 us | 0.202 us |   2.6245 |  0.0305 |     - |  16.24 KB |


|        Method |        Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |------------:|---------:|---------:|---------:|--------:|------:|----------:|
| Introspection | 1,134.23 us | 5.651 us | 4.719 us | 117.1875 | 35.1563 |     - | 729.25 KB |
|          Hero |    24.48 us | 0.103 us | 0.097 us |   2.6245 |  0.0305 |     - |  16.24 KB |

```

Shane32 commented:
What this really shows is how much dictionaries and allocations and stuff can impact performance.  I just know we can get a LOT more performance out if we work at it together.  Even here in the validation stuff, which should only really be a small portion of the execution time, we have a 10% overall performance difference between two different approaches.

Shane32 commented:
Can you imagine how much this validation code would benefit from ArrayPool ?  I discouraged use of it for the ResolveFieldContext code a while back, but here I think it would work wonders.  If you're open to adding a dependency, I'd like to try implementing it (separate PR) and run some performance tests to see what effect it would have.

Shane32 commented:
I have read research that says that ArrayPool can actually hurt performance for very small arrays, as the GC is very efficient, so maybe it would not be that useful.

sungam3r commented:
What do we do with our two PRs?

Shane32 commented:
Well, first I'd like to read through your whole PR.  I have not done so yet.  I assume that you agree that performance is what we are striving for, and the improvements (as it stands in this PR) doesn't meet that goal.

Shane32 commented:
I'd also like you to review my last commit.  Basically, I removed `EnterLeaveListener` and `LightweightCache` and replaced them.

Shane32 commented:
I've not written xml comments yet, but we need to be 'on the same page' before I spend a lot more time on this.

sungam3r commented:
> I'd also like you to review my last commit. Basically, I removed EnterLeaveListener and LightweightCache and replaced them.

That's it. Now the difference lies in the **additional** details. My main goal was to make absolutely all validation rules static (stateless). Now I can further improve the results.

Shane32 commented:
FYI `EnterLeaveListener` is an allocation itself and creates a `List` allocation, and the list creates an array allocation, and then items are added one at a time.  If the array is too small, it would have to copy the array causing another allocation.  And if it's too large then memory is wasted.  It is now replaced with a much simpler class that uses an array passed to it (already sized correctly), eliminating at least one allocation plus some CPU cycles.

Shane32 commented:
> > I'd also like you to review my last commit. Basically, I removed EnterLeaveListener and LightweightCache and replaced them.
> 
> That's it. Now the difference lies in the **additional** details. My main goal was to make absolutely all validation rules static (stateless). Now I can further improve the results.

Yes but as it stands now your performance is much worse than it started at.  I don't think you should spend time attempting to improve a codebase that is performing poorer than it was before.  100% static validation rules are only beneficial when they improve performance, and it does not do so (currently).

I have not introduced any code that runs slower than previously.

sungam3r commented:
> FYI EnterLeaveListener is an allocation itself and creates a List allocation, and the list creates an array allocation, and then items are added one at a time. If the array is too small, it would have to copy the array causing another allocation. And if it's too large then memory is wasted. It is now replaced with a much simpler class that uses an array passed to it (already sized correctly), eliminating at least one allocation plus some CPU cycles.

sungam3r commented:
> I don't think you should spend time attempting to improve a codebase that is performing poorer than it was before. 

Conceptually, I did everything correctly, so I'm sure that somewhere I was just mistaken. If it's all about the dictionary, then it can be solved.

sungam3r commented:
I have localized possible problem areas. I will work on this PR tomorrow.

Shane32 commented on src/GraphQL/Validation/BasicVisitor.cs:
```diff
@@ -38,7 +38,8 @@ public void Visit(INode node)
             for (int i = 0; i < _visitors.Count; i++)
             {
-                _visitors[i].Enter(node);
+                if (_visitors[i].ShouldRunOn(context))
```
Within the document validator, you're filtering by ShouldRunOn, so it's not needed here, correct?

sungam3r replied:
I know. But theoretically this class can be used on its own.

Shane32 commented on src/GraphQL/Validation/Rules/NoFragmentCycles.cs:
```diff
@@ -13,33 +13,41 @@ namespace GraphQL.Validation.Rules
-        }
+                });
+            },
+            shouldRun: context => context.Document.Fragments.Count > 0
```
Nice 👍

sungam3r replied:
It's a pity that not all the rules are so simple to bypass. Although the possibility of such a check already shows us how expensive the current implementation of the rules is.

Shane32 commented on src/GraphQL/Validation/Rules/UniqueFragmentNames.cs:
```diff
@@ -34,7 +32,12 @@ public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        }
+                });
+            },
+            shouldRun: context => context.Document.Fragments.Count >= 2).ToTask();
```
👍 
I like the idea of short circuiting some of these rules.

Shane32 commented on src/GraphQL/Validation/ValidationContext.cs:
```diff
@@ -20,6 +21,11 @@ public class ValidationContext : IProvideUserContext
+        /// <summary>
+        /// Allows validation rules store their specific data during validation.
+        /// </summary>
+        private readonly ConcurrentDictionary<object, object> _validationRuleLocalContext = new ConcurrentDictionary<object, object>();
```
Since `Get<TValidationRule, TResult>` and `Set<>` are `internal` and not public, there is no need for a concurrent dictionary.

sungam3r replied:
Why? Rules are executed in parallel by `DocumentValidator`.

Shane32 replied:
No they are not.  They "asynchronously" retrieve the `INodeVisitor` instances, and then run them synchronously in the `BasicVisitor`.  But since none of the rules have an `await` in them, they will all return the `INodeVisitor` instances synchronously also.  It's 100% synchronous with the default rule set (whether its the old code in develop, or any of our branches).

The only thing that could be asynchronous is if there was a custom validation rule that, _before_ returning the `INodeVisitor`, awaited another method -- that portion of that one validation rule's code would be asynchronous.  But that is not part of the core rule set that has access to this dictionary.

sungam3r replied:
🤦 🤦 🤦  I remember thinking about it, but then something distracted me and I forgot.

Shane32 commented on src/GraphQL/Validation/DocumentValidator.cs:
```diff
@@ -81,24 +83,34 @@ public class DocumentValidator : IDocumentValidator
+            else
+            {
+                visitors = (await Task.WhenAll(rules.Select(x => x.ValidateAsync(context)))).Where(v => v.ShouldRunOn(context)).ToList();
+                visitors.Insert(0, context.TypeInfo);
```
I don't think this "optimization" is worth the benefit, if there even is one.  Typically the rule set will inherit from `IList<IValidationRule>` and when that's the case:

1. The size of allocated list is known either way.
2. There is no need for LINQ either way.
3. There are no runtime allocations for a simple `await` on a completed task.  Since the core rules use `Task.FromResult`, there would normally be no allocations with a simple `await`.  I'm not positive but it's likely that there are no allocations for `Task.WhenAll` when all of the tasks have completed.  If this was not the case, we could optimize this like I did in the execution strategy (which would work for any rule set), but if MS did their job with `WhenAll` then it won't be any faster no matter what.

Shane32 replied:
And this code only runs once per document being validated.  Eliminating one or two allocations isn't going to have a measurable impact, even at the nanosecond level.

sungam3r replied:
This can be viewed and measured separately.

Shane32 commented:
> > FYI EnterLeaveListener is an allocation itself and creates a List allocation, and the list creates an array allocation, and then items are added one at a time. If the array is too small, it would have to copy the array causing another allocation. And if it's too large then memory is wasted. It is now replaced with a much simpler class that uses an array passed to it (already sized correctly), eliminating at least one allocation plus some CPU cycles.
> 
> I understand but in case of statics it doesn't matter because it only happens once.

I would normally agree, but my performance testing seemed to indicate that the `EnterLeaveListener` still had slower runtime performance than my `NodeListeners` class even with a static instance.  I figured it was because the `foreach` iterates directly over an array, and there are less allocations or checks or something -- but it could be due to another reason.  It's probably normally a miniscule difference, but in this case there are LOT of executions of the `EnterLeaveListener.Enter` and `Leave` method, so every tiny bit helps.

sungam3r commented:
Rejected as already done in other PRs.
